### PR TITLE
Refactor `istate_charge.cpp` using new DensityMatrix class

### LINKED ISSUE #4090: Refactoring & Optimization: Streamlining Density Matrix Management with respect to `Local Orbital Charge`
### Background
The `Local Orbital Charge` class, a significant component within our LCAO-related codebase, is presently marked by extensive global class usage. This widespread reliance on global classes has notably impaired the code's maintainability, making it challenging to integrate unit tests and new features without escalating the cost of maintenance.

### What to do
Our primary objective is to begin disentangling the dependencies around the density matrix (`DM`) representation within the `Local Orbital Charge` class. This task serves as the initial step in a broader plan to enhance code modularity and reduce global dependencies.

The current approach involves direct manipulation of `DM` across various parts of the project, which has contributed to tightly coupled code and complex maintenance challenges. To address these issues, the immediate plan includes segregating functions directly dependent on `DM` into a dedicated file within the `module_io` directory. Subsequently, the old `DM` will be replaced progressively with the newly structured density matrix format (final goal).

This refactoring is not merely a reorganization of existing code but a strategic move towards a more modular and maintainable structure. By minimizing direct interactions with global variables and enhancing the clarity of data flows within our software, we aim to significantly improve the robustness and testability of our codebase.

Furthermore, functions like `output_dm` and `output_dm1` that manage density matrix outputs will be scrutinized for their current roles and differences to ensure that their functionalities align with the new architectural standards.

Details regarding the phased implementation of these changes will follow. This initial issue aims to set the groundwork for subsequent updates that will further encapsulate and isolate functionality, paving the way for a cleaner, more efficient codebase.

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [X] Identify the specific code file or section with the code quality issue.
- [X] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified issue.
- [ ] Refactor the code to improve code quality, following the suggested solution.
- [ ] Ensure the refactored code adheres to the project's coding standards.
- [ ] Test the refactored code to ensure it functions as expected.
- [ ] Update any relevant documentation, if necessary.
- [ ] Submit a pull request with the refactored code and a description of the changes made.

### PR DESCRIPTION
**Pull request by AsTonyshment**
### Linked Issue
Fix #4090. 

### What's changed?
Refactor `istate_charge.cpp` using new `DensityMatrix` class. But actually, we cannot get rid of `LOC` completely at this moment, so this issue will be reconsidered in the near future.

### CODE REVIEW

File: source/module_io/istate_charge.cpp [hunk 1]
```diff
@@ -195,20 +198,32 @@ void IState_Charge::begin(Gint_Gamma& gg,
...
-            // (1) calculate the density matrix for a partuclar
-            // band, whenever it is occupied or not.
+            // (1) calculate the density matrix for a partuclar band, whenever it is occupied or not.
+
+            // Using new density matrix inplementation
+            elecstate::DensityMatrix<double, double> DM(this->loc->ParaV, nspin);
 
 #ifdef __MPI
-            this->idmatrix(ib, nspin, nelec, nlocal, wg);
+            this->idmatrix(ib, nspin, nelec, nlocal, wg, DM);
```
Comment (mohanchen) [inline source/module_io/istate_charge.cpp:209]: else if not MPI, what will happen?

File: source/module_io/istate_charge.cpp [hunk 2]
```diff
@@ -195,20 +198,32 @@ void IState_Charge::begin(Gint_Gamma& gg,
...
             for (int is = 0; is < nspin; ++is)
             {
                 ModuleBase::GlobalFunc::ZEROS(rho[is], rhopw_nrxx);
             }
 
-            // (3) calculate charge density for a particular
-            // band.
+            // (3) calculate charge density for a particular band.
+
+            DM.init_DMR(&GlobalC::GridD, &GlobalC::ucell);
```
Comment (mohanchen) [inline source/module_io/istate_charge.cpp]: try not using GlobalC

File: source/module_io/istate_charge.h [hunk 3]
```diff
@@ -68,8 +70,14 @@ class IState_Charge
      * @param nelec Total number of electrons.
      * @param nlocal Number of local orbitals.
      * @param wg Weight matrix for bands and spins.
+     * @param DM Density matrix to be calculated.
      */
-    void idmatrix(const int& ib, const int nspin, const double nelec, const int nlocal, const ModuleBase::matrix& wg);
+    void idmatrix(const int& ib,
+                  const int nspin,
+                  const double nelec,
```
Comment (mohanchen) [inline source/module_io/istate_charge.h]: add & before nelec

File: source/module_io/istate_charge.cpp (hunk 1 above)
Comment (AsTonyshment) [reply source/module_io/istate_charge.cpp:209]: Currently there is no non-MPI version for `idmatrix`. I have added a WARNING_QUIT in else branch to handle this. See [PR#712](https://github.com/deepmodeling/abacus-develop/pull/712).

**Review by mohanchen — APPROVED** [review]

**Pull request merged**